kf_lib_data_ingest/validation/hierarchy.py

```diff
@@ -0,0 +1,97 @@
+from math import inf
+
+from graph import Graph  # https://github.com/root-11/graph-theory
+from kf_lib_data_ingest.common.concept_schema import CONCEPT as C
+from kf_lib_data_ingest.validation.relations import RELATIONS as R
+
+
+def get_full_hierarchy(H=None):
+    HIERARCHY = H if H is not None else DH
+
+    # faster lookup than HIERARCHY.nodes()
+    HIERARCHY_DIRECT = {n: HIERARCHY.nodes(n) for n in HIERARCHY.nodes()}
+
+    # used for listing stable node order from top to bottom
+    _reverse = Graph()
+    for a, b, _ in HIERARCHY.edges():
+        _reverse.add_edge(b, a)
+
+    top = None
+    for n, v in HIERARCHY_DIRECT.items():
+        if not v:
+            top = n
+            break
+    HIERARCHY_ORDER = list(_reverse.breadth_first_walk(top))
```

Comment:
There could be multiple tops. This picks one to start from, but yours would miss them.

Reply:
HIERARCHY_ORDER is only used for display, so I may just get rid of it entirely. If anything, I think the reverse graph will need to be made bidirectional instead in case there are multiple top nodes.